add-bookings

ilya-noize opened the pull request:
Проект ShareIt. Техническое задание. Часть 2

- добавить работу с базой данных в уже реализованную часть проекта;
- дать пользователям возможность бронировать вещи и оставлять отзывы на взятые в аренду вещи;
- показывать списки предметов в аренду и взятых в аренду для арендодателей и арендаторов.

ilya-noize committed:
Maven: data jpa
Booking: entity, mapper, repository, dto, states(enum), status(enum), controller;
Remastering Users, Items to JPA;

todo: fix 4 errors!

ilya-noize committed:
Unused classes have been removed;

ilya-noize committed:
Troubleshooting startup errors

ilya-noize committed:
Troubleshooting tests issues

ilya-noize committed:
Troubleshooting tests issues

ilya-noize committed:
It's almost there...

ilya-noize committed:
Finally, this shit is over!!!11111

ilya-noize committed:
Finally, this shit is over!!!11111

ilya-noize committed:
Merge branch 'add-bookings' of https://github.com/ilya-noize/java-shareit into add-bookings

TyutterinYakov commented on src/main/java/ru/practicum/shareit/booking/api/dto/BookingDto.java:
```diff
@@ -0,0 +1,29 @@
+    private Long id;
+
+    @FutureOrPresent
+    private LocalDateTime start;
```
Не забываем также использовать аннотацию `@NotNull`, так как `@FutureOrPresent` допускает это значение за корректное)

TyutterinYakov commented on src/main/java/ru/practicum/shareit/booking/api/dto/BookingDto.java:
```diff
@@ -0,0 +1,29 @@
+    @FutureOrPresent
+    private LocalDateTime start;
+
+    @Future
```
Не забываем также использовать аннотацию `@NotNull`, так как `@Future` допускает это значение за корректное)

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/api/dto/ItemDto.java:
```diff
@@ -20,9 +26,9 @@
+@Getter
+@Setter
+@Builder()
 public class ItemDto {
```
Хорошо было бы разбить класс-DTO на несколько, так как, например, при создании нам требуются лишь поля с описанием, названием и доступностью) Остальные же не требуются. То есть один класс будет использоваться для представления тела запроса, а другой - для представления тела ответа) Таким образом мы будем работать только с теми значениями, что требуются в определенной ситуации

TyutterinYakov replied:
Хорошо было бы ограничить размеры текстовых полей на основе тех ограничений, которые установлены на уровне таблицы в БД, чтобы заранее откинуть некорректные запросы) Для ограничения количества символов можно использовать `@Size`. Эта аннотация удобна тем, что допускает null, за счет чего при обновлении по значению мы сможем понять, было передано значение или нет)

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/api/dto/ItemSimpleDto.java:
```diff
@@ -0,0 +1,41 @@
+@Data
+@Builder
+@AllArgsConstructor
+public class ItemSimpleDto {
```
Действительно ли в этом классе требуются аннотации валидации?) Если я не ошибаюсь, то он выступает только в качестве возвращаемого значения, и оно никак не валидируется

TyutterinYakov commented on src/main/java/ru/practicum/shareit/booking/api/BookingServiceImpl.java:
```diff
@@ -0,0 +1,316 @@
+        LocalDateTime end = dto.getEnd();
+
+
+        if (start.equals(end)) {
```
(предложение)Эту валидацию можно было бы реализовать с помощью аннотации валидации над классом, чтобы поддерживать декларативный подход https://devcolibri.com/spring-mvc-кастомная-аннотация-для-валидации/
Кстати, если будет интересно то, как реализовать кастомную аннотацию валидации на уровне класса `BookingDto` - можешь чекнуть тут) https://github.com/TyutterinYakov/CustomValidationClassLevel Я написал небольшой примерчик, чтобы было проще разобраться в теме)

TyutterinYakov commented on src/main/java/ru/practicum/shareit/user/entity/User.java:
Для `UserDto`
Хорошо было бы ограничить размеры текстовых полей на основе тех ограничений, которые установлены на уровне таблицы в БД, чтобы заранее откинуть некорректные запросы) Для ограничения количества символов можно использовать `@Size`. Эта аннотация удобна тем, что допускает null, за счет чего при обновлении по значению мы сможем понять, было передано значение или нет)

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/comment/api/dto/CommentDtoSource.java:
```diff
@@ -0,0 +1,21 @@
+@AllArgsConstructor
+@Getter
+@Setter
+public class CommentDtoSource {
```
Хорошо было бы ограничить размеры текстовых полей на основе тех ограничений, которые установлены на уровне таблицы в БД, чтобы заранее откинуть некорректные запросы) Для ограничения количества символов можно использовать `@Size`. Эта аннотация удобна тем, что допускает null, за счет чего при обновлении по значению мы сможем понять, было передано значение или нет)

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/comment/api/dto/CommentDtoSource.java:
```diff
@@ -0,0 +1,21 @@
+@AllArgsConstructor
+@Getter
+@Setter
+public class CommentDtoSource {
```
Хорошо было бы разбить класс-DTO на несколько, так как, например, при создании нам требуются лишь поле с текстом) Остальные же не требуются. То есть один класс будет использоваться для представления тела запроса, а другой - для представления тела ответа) Таким образом мы будем работать только с теми значениями, что требуются в определенной ситуации

ilya-noize replied:
Разделён.
Только другие поля подтянуты по аналогии с другими решениями и прекрасно работало.
Так и осталось )

TyutterinYakov replied:
Да, но они не требуются при создании, так как от пользователя мы ожидаем лишь текст комментария) Остальные же следует исключить из этого DTO

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/comment/api/dto/CommentDtoRecord.java:
```diff
@@ -0,0 +1,20 @@
+@AllArgsConstructor
+@Getter
+@Setter
+public class CommentDtoRecord {
```
Действительно ли в этом классе требуются аннотации валидации?) Если я не ошибаюсь, то он выступает только в качестве возвращаемого значения, и оно никак не валидируется

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/api/ItemServiceImpl.java:
```diff
@@ -3,184 +3,409 @@
+                                userId, itemId)));
+
+        bookings.stream()
+                .filter(booking -> booking.getEnd().isBefore(LocalDateTime.now()))
```
Фильтрацию лучше произвести на уровне запроса к БД, чтобы не производить получение лишних данных)

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/api/ItemServiceImpl.java:
```diff
@@ -3,184 +3,409 @@
+                .orElseThrow(() -> new NotFoundException(
+                        format(ITEM_WITH_ID_NOT_EXIST, itemId)));
+        List<Booking> bookings = bookingRepository
+                .findAllByItemIdAndBookerIdAndStatus(
```
(предложение) Вместо `findBy` можно было бы использовать `exists`, метод репозитория вернет просто значение `true` или `false` в зависимости от того, было найдено значение или нет) Таким образом мы не будет производить ненужную десериализацию, так как нам важно проверить лишь факт наличия)

TyutterinYakov commented on src/main/java/ru/practicum/shareit/booking/api/repository/BookingRepository.java:
```diff
@@ -0,0 +1,91 @@
+            BookingStatus status);
+
+    // - - - - - - - - - - - - - - - - - - APPROVED, WAITING BOOKER
+    @Query("select b from Booking b where b.booker.id = ?1 and b.status = ?2 order by b.start DESC")
```
Можно было бы оставить возможность выбора сортировки, чтобы в случае, когда нам потребуется другой порядок - не писать еще один метод) Для этого в аргументах метода следует так же ожидать `Sort sort`.
Объект этого класса создается подобным образом: `Sort.by(Sort.Direction.DESC, "start»);`
А `order by b.start DESC ` следовательно нужно будет убрать)

ilya-noize replied:
Считаю это излишним.
Получение бронирований в обратном порядке (от ранних к поздним) не подразумевается в тз.
Кроме этого, почти везде идёт именно эта сортировка.
Сигнатура метода выглядит проще.

TyutterinYakov replied:
Это как правка на будущее) При внесении этих изменений, в следующем спринте, при добавлении пагинации, правки будут легче) (можно не исправлять)

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/api/ItemServiceImpl.java:
```diff
@@ -3,184 +3,409 @@
+        List<Integer> itemIds = itemsDto.stream()
+                .map(ItemDto::getId).collect(toList());
+        List<Booking> bookings = bookingRepository
+                .findAllByItem_Owner_IdOrderByStartDesc(userId);
```
Получение лучше выполнить по набору идентификаторов вещей, которые мы получаем, а не по идентификатору владельца вещей) Так как в будущем мы добавим пагинацию, а значит набор вещей будет ограничен, а получение лишних бронирований - будет не очень эффективно)

ilya-noize replied:
пагонация от page ) ok

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/api/ItemServiceImpl.java:
```diff
@@ -3,184 +3,409 @@
+        List<BookingToItemDto> bookingsToItemDto = bookings.stream()
+                .map(bookingMapper::toItemDto)
+                .collect(Collectors.toList());
+        List<CommentEntity> comments = commentRepository
```
Комментарии, бронирования лучше преобразовать к мапе, где ключом будет идентификатор вещи, а значением коллекция бронирований/комментариев) Так как иначе, логика ниже не будет эффективной, так как сейчас сложность алгоритма составляет O(n^2), так как мы в цикле проходимся каждый раз по одной и той же коллекции бронирований и вещей)

Пример для комментариев:
```
		Map<Item, List<Comment>> comments = commentRepository.findByItemIn(items, Sort.by(DESC, "created"))
				.stream()
				.collect(groupingBy(Comment::getItem, toList()));
```

При получении из мапы затем удобно будет использовать метод `getOrDefault`

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/api/ItemServiceImpl.java:
```diff
@@ -3,184 +3,409 @@
+            Integer itemId) {
+
+        return comments.stream()
+                .filter(comment ->
```
Условие более не потребуется)

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/api/ItemServiceImpl.java:
```diff
@@ -3,184 +3,409 @@
+                    return itemId.equals(itemBooking)
+                            && start.isBefore(LocalDateTime.now());
+                })
+                .reduce((a, b) -> a)
```
В данном случае, для получения первого бронирования из списка можно было бы использовать `findFirst`

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/api/ItemServiceImpl.java:
```diff
@@ -3,184 +3,409 @@
+                    Integer itemBooking = booking.getItemId();
+                    LocalDateTime start = booking.getStart();
+                    return itemId.equals(itemBooking)
+                            && start.isBefore(LocalDateTime.now());
```
Лучше использовать условие вида:
`!start.isAfter(now)`, так как время старта последнего бронирования может быть равным текущему)

TyutterinYakov replied:
Текущую дату лучше передавать в параметрах метода из того метода, откуда он вызывается) Чтобы фильтрация всех бронирований осуществлялась по одному моменту времени)

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/api/ItemServiceImpl.java:
```diff
@@ -3,184 +3,409 @@
+                .filter(booking -> {
+                    Integer itemBooking = booking.getItemId();
+                    LocalDateTime start = booking.getStart();
+                    return itemId.equals(itemBooking)
```
Условие равенства идентификаторов более не потребуется)

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/api/ItemServiceImpl.java:
```diff
@@ -3,184 +3,409 @@
+                .filter(booking -> {
+                    Integer itemBooking = booking.getItemId();
+                    LocalDateTime start = booking.getStart();
+                    return itemId.equals(itemBooking)
```
Условие равенства идентификаторов более не потребуется)

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/api/ItemServiceImpl.java:
```diff
@@ -3,184 +3,409 @@
+                    Integer itemBooking = booking.getItemId();
+                    LocalDateTime start = booking.getStart();
+                    return itemId.equals(itemBooking)
+                            && start.isAfter(LocalDateTime.now());
```
Текущую дату лучше передавать в параметрах метода из того метода, откуда он вызывается) Чтобы фильтрация всех бронирований осуществлялась по одному моменту времени)

TyutterinYakov reviewed (commented):
Добрый вечер! В целом - хорошо) Отметил то, что следует исправить и то, что можно сделать лучше) Хорошего вечера! Если возникнут вопросы - пиши в пачку, буду рад помочь🌃

ilya-noize committed:
NotNull fields for Validate

ilya-noize committed:
Size fields for Validate

ilya-noize committed:
Reviewer's comments #1

ilya-noize committed:
Reviewer's comments #2
ItemSimpleDto in create, update;
ItemDto just a class with data/

ilya-noize committed:
Reviewer's comments #3
itemService - getAll - remaster by map;

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/api/ItemController.java:
```diff
@@ -4,62 +4,78 @@
     public ItemDto create(
-            @RequestHeader(headerUserId)
+            @RequestHeader(HEADER_USER_ID)
             @NotNull(groups = Create.class)
```
Аннотация `@NotNull` не имеет смысла, так как заголовок запроса является обязательным, так как в `@RequestHeader` не указано обратное - `defaultValue = ".."` или `required = false`

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/api/ItemController.java:
```diff
@@ -4,62 +4,78 @@
-            @RequestBody @Validated(Update.class) ItemDto itemDto) {
+            @RequestHeader(HEADER_USER_ID) Integer userId,
+            @PathVariable(name = "id") Integer itemId,
+            @RequestBody ItemSimpleDto itemDto) {
```
Не забываем добавить `@Validated(Update.class)`, чтобы аннотации валидации для обновления стали учитываться)

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/api/dto/ItemSimpleDto.java:
```diff
@@ -0,0 +1,41 @@
+    private String name;
+
+    @NotBlank(groups = {Create.class})
+    @Size(max = 512)
```
Не забываем указать две группы валидации, так как она должна учитываться при создании и при обновлении)

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/api/dto/ItemSimpleDto.java:
```diff
@@ -0,0 +1,41 @@
+    private Integer id;
+
+    @NotBlank(groups = {Create.class})
+    @Size(max = 255)
```
Не забываем указать две группы валидации, так как она должна учитываться при создании и при обновлении)

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/comment/api/dto/CommentDtoSource.java:
```diff
@@ -0,0 +1,23 @@
+public class CommentDtoSource {
+    private final Integer id;
+    @NotBlank(groups = {Create.class})
+    @Size(max = 2048)
```
Не забываем указать `{Create.class}`, иначе аннотация валидации не будет учтена, так как мы используем `@Validated` с указанием группы

TyutterinYakov commented on src/main/java/ru/practicum/shareit/user/api/dto/UserDto.java:
```diff
@@ -5,19 +5,20 @@
+
     @Email(groups = {Create.class, Update.class})
     @NotEmpty(groups = {Create.class})
+    @Size(max = 255)
```
Аналогично, не забываем указать две группы)

TyutterinYakov commented on src/main/java/ru/practicum/shareit/user/api/dto/UserDto.java:
```diff
@@ -5,19 +5,20 @@
     private String email;
+
     @NotBlank(groups = {Create.class})
+    @Size(max = 255)
```
Аналогично, не забываем указать две группы)

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/api/ItemServiceImpl.java:
```diff
@@ -2,185 +2,406 @@
+                                userId, itemId)));
+
+        bookings.stream()
+                .filter(booking -> booking.getEnd().isBefore(LocalDateTime.now()))
```
Фильтрацию лучше произвести на уровне запроса к БД, чтобы не производить получение лишних данных)

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/api/ItemServiceImpl.java:
```diff
@@ -2,185 +2,406 @@
-                .map(mapper::toDto)
+
+        return itemRepository
+                .findByNameContainsIgnoreCaseOrDescriptionContainsIgnoreCaseAndAvailableTrueOrderByIdAsc(
```
Корректный ли получается запрос?)
Так как во втором условии используется AND, это приводит к запросу вида:
`select * from items where name LIKE %?1% OR (description LIKE %?1% AND available = true)`
А должно быть:
`select * from items where (name LIKE %?1% OR description LIKE %?1%) AND available = true`

ilya-noize replied:
Прошло pipeline )
Про оператор "скобка" я не слышал. Он существует? )

TyutterinYakov replied:
Нет, но здесь следует использовать запрос, написанный в `@Query`, чтобы избежать этой проблемы)

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/api/ItemServiceImpl.java:
```diff
@@ -2,185 +2,406 @@
+        }
+
+        return comments.stream()
+                .filter(comment ->
```
`filter`  более не требуется, так как выше мы произвели разбиение на мапу) А в метод передаем лишь комментарии конкретной вещи)

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/api/ItemServiceImpl.java:
```diff
@@ -2,185 +2,406 @@
                 .collect(toList());
+
+        Map<Integer, BookingToItemDto> lastBookingStorage = bookingRepository
+                .findFirstByItem_IdInAndStartLessThanEqualAndStatus(
```
`First` не сработает, в данном случае, так как будет получено лишь бронирование с большим временем старта среди всех бронирований)

ilya-noize replied:
Тесты проходят ) Получается что сработал.
А нужно сопоставлять itemId со всеми бронированиями этого item? Если так, то много лишних данных буду получать из БД.

ilya-noize replied:
При удалении First не проходит тест https://web.postman.co/workspace/filmorate~f2060cd4-1e7c-430c-b2eb-996d7dcc6700/request/28278571-0ae777c7-0670-482f-8af2-c78168d06d50

Или тест сломан, или ... какие-то нюансы с JPQL при создании списка бронирований не учтены.
Хотя запрашиваю как раз список. Может нужно сопоставлять список с itemId, а не одно конкретное бронирование ближайшее к `сейчас`?

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/api/ItemServiceImpl.java:
```diff
@@ -2,185 +2,406 @@
+                .collect(toMap(BookingToItemDto::getItemId, Function.identity()));
+
+        Map<Integer, BookingToItemDto> nextBookingStorage = bookingRepository
+                .findFirstByItem_IdInAndStartAfterAndStatus(
```
`First` не сработает, в данном случае, так как будет получено лишь одно бронирование с меньшим временем старта среди всех бронирований)

TyutterinYakov reviewed (commented):
Доброе утро :) Отметил оставшиеся недочеты) Хорошего дня! Если возникнут вопросы - пиши в пачку, буду рад помочь🌅

ilya-noize committed:
Reviewer's comments #4
Validate groups

ilya-noize committed:
Reviewer's comments #5
JPQL

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/api/ItemServiceImpl.java:
```diff
@@ -2,185 +2,402 @@
+                                userId, itemId)));
+
+        bookings.stream()
+                .filter(booking -> booking.getEnd().isBefore(LocalDateTime.now()))
```
Фильтрацию лучше произвести на уровне запроса к БД, чтобы не производить получение лишних данных)

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/comment/api/dto/CommentDtoSource.java:
```diff
@@ -0,0 +1,23 @@
+@AllArgsConstructor
+@Getter
+@Setter
+public class CommentDtoSource {
```
<img width="781" alt="image" src="https://github.com/ilya-noize/java-shareit/assets/45480231/fc2019a5-a0ba-4ccd-9c7c-8a254f678d44">
Также хорошо было бы исправить)

ilya-noize replied:
`git reset --hard`
Помню что поправил, но после ресета не поправил.

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/api/ItemServiceImpl.java:
```diff
@@ -2,185 +2,402 @@
-                .map(mapper::toDto)
+
+        return itemRepository
+                .findByNameContainsIgnoreCaseOrDescriptionContainsIgnoreCaseAndAvailableTrueOrderByIdAsc(
```
Ответил в прошлом замечании) Подправь, пожалуйста)

ilya-noize replied:
```
    @Query("select i from Item i " +
            "where ( " +
            "upper(i.name) like upper(concat('%', :search, '%')) or upper(i.description) like upper(concat('%', :search, '%')) " +
            ") and i.available = true " +
            "order by i.id")
```

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/api/ItemServiceImpl.java:
```diff
@@ -2,185 +2,402 @@
                 .collect(toList());
+
+        Map<Integer, BookingToItemDto> lastBookingStorage = bookingRepository
+                .findFirstByItem_IdInAndStartLessThanEqualAndStatus(
```
Нет, тест не сломан) Возможно, не совсем корректно происходит преобразование списка к мапе, так как используется `Function.identity()`, хотя мы должны понимать, что к одной вещи может относиться более одного бронирования, подходящие под текущие условия поиска в БД)
Вот тут описано, как можно поступить:
https://www.baeldung.com/java-duplicate-keys-when-producing-map-using-stream#using-thetomapmethod-and-handling-the-duplicated-keys

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/api/ItemServiceImpl.java:
```diff
@@ -2,185 +2,402 @@
+                .collect(toMap(BookingToItemDto::getItemId, Function.identity()));
+
+        Map<Integer, BookingToItemDto> nextBookingStorage = bookingRepository
+                .findFirstByItem_IdInAndStartAfterAndStatus(
```
ЗДесь также мы должны исключить `First`

TyutterinYakov reviewed (commented):
Добрый вечер! Отметил оставшиеся критичные недочеты) Хорошего вечера! Если возникнут вопросы по ревью - пиши в пачку, буду рад помочь🌃

ilya-noize committed:
Reviewer's comments #6
JPQL - BookingRepository, ItemRepository
CommentMapper - add toEntity
CommentSimpleDto - remove id;
ItemController - send CommentSimpleDto to service;

ilya-noize committed:
Reviewer's comments #6
fixed {"authorName" : "updateName"}

TyutterinYakov reviewed (approved):
Добрый вечер! Поздравляю с успешной сдачей спринта :) Сил и терпения на этом сложном, но интересном пути! Успехов в обучении и прекрасного вечера!🌃🎇

ilya-noize reviewed (commented):
Применил на практике зависимость **ManyToOne**.
Научился работать со сложными **DTO**-классами.
Применил на практике **JPQL**.
Узнал как создать преобразовать **List** в **Map** c помощью **stream.collect(toMap(...));** другим способом (не таким как в _filmorate_)
Это была тяжёлая долгая работа по всем фронтам )
Если не изменяет память, трижды всё переписывалось с начала.
Что-то забывалось, что-то не совсем корректно работало.
Но я её всё таки сделал.
И благодаря качественному ревью можно спокойно продолжить разработку дальше.

Благодарю за отличное ревью моей работы, @TyutterinYakov )
